elfpy/utils/apeworx_integrations.py

```diff
@@ -198,37 +195,55 @@ def get_wallet_from_onchain_trade_info(
-        )
+        mint_time = maturity - elfpy.SECONDS_IN_YEAR
+        log_and_show(f" => {asset_type}({asset_prefix}) maturity={maturity} mint_time={mint_time}")
+
+        # verify our calculation against the onchain balance
+        on_chain_balance = hyperdrive_contract.balanceOf(position_id, address_)
+        if abs(balance - on_chain_balance) > elfpy.MAXIMUM_BALANCE_MISMATCH_IN_WEI:
+            raise ValueError(
+                f"events {balance=} and {on_chain_balance=} disagree by "
+                f"more than {elfpy.MAXIMUM_BALANCE_MISMATCH_IN_WEI} wei for {address_}"
+            )
+        log_and_show(f" => calculated balance = on_chain = {fmt(balance)}")
 
         # check if there's an outstanding balance
-        if balance != 0 or hyperdrive_contract.balanceOf(position_id, address_) != 0:
-            # loop across all the positions owned by this wallet
-            for specific_trade in trades_in_position.index[trades_in_position]:
-                if asset_type == "SHORT":
-                    open_share_price = info.share_price[info.trades.loc[specific_trade, "block_number"]]
-                    wallet.shorts |= {
-                        maturity - elfpy.SECONDS_IN_YEAR: Short(balance=balance, open_share_price=open_share_price)
-                    }
-                elif asset_type == "LONG":
-                    wallet.longs |= {maturity - elfpy.SECONDS_IN_YEAR: Long(balance=balance)}
-                elif asset_type == "LP":
-                    wallet.lp_tokens += balance
+        if balance != 0 or on_chain_balance != 0:
+            if asset_type == "SHORT":
+                # loop across all the positions owned by this wallet
+                sum_product_of_open_share_price_and_value, sum_value = 0, 0
+                for specific_trade in trades_in_position.index[trades_in_position]:
+                    value = info.trades.loc[specific_trade, "value"]
+                    value *= -1 if info.trades.loc[specific_trade, "from"] == address_ else 1
+                    sum_value += value
+                    sum_product_of_open_share_price_and_value += (
+                        value * info.share_price[info.trades.loc[specific_trade, "block_number"]]
+                    )
+                open_share_price = sum_product_of_open_share_price_and_value / sum_value
+                assert (
+                    abs(balance - sum_value) <= elfpy.MAXIMUM_BALANCE_MISMATCH_IN_WEI
+                ), "weighted average open share price calculation is wrong"
+                print(f"calculated weighted average open share price of {open_share_price}")
+                new_record = {mint_time: Short(balance=balance, open_share_price=open_share_price)}
+                wallet.shorts |= new_record
+                print(f"storing in wallet as {new_record=}")
+            elif asset_type == "LONG":
+                new_record = {mint_time: Long(balance=balance)}
+                wallet.longs |= new_record
```

Comment:
`|=` performs different operations for different types. In this case, it's a dictionary update. But a new programmer looking over the code might not recall that `wallet.longs` is a dictionary. For this reason, I think it is better to use the more explicit `update` syntax.

Reply:
ya apparently its behavior on sets is quite different. it was an operator there before they added it to dicts.